File: Evaluator.py
import math
import cmath
from time import monotonic_ns

trigs = {
  'sin': lambda x: cmath.sin(x) if isinstance(x, complex) else math.sin(x),
  'cos': lambda x: cmath.cos(x) if isinstance(x, complex) else math.cos(x),
  'tan': lambda x: cmath.sin(x) / cmath.cos(x) if isinstance(x, complex) else math.tan(x),
  'csc': lambda x: 1 / (cmath.sin(x) if isinstance(x, complex) else math.sin(x)),
  'sec': lambda x: 1 / (cmath.cos(x) if isinstance(x, complex) else math.cos(x)),
  'cot': lambda x: 1 / (cmath.tan(x) if isinstance(x, complex) else math.tan(x)),
  'asin': lambda x: cmath.asin(x) if isinstance(x, complex) else math.asin(x),
  'acos': lambda x: cmath.acos(x) if isinstance(x, complex) else math.acos(x),
  'atan': lambda x: cmath.atan(x) if isinstance(x, complex) else math.atan(x)
}

operate = {
  0: lambda x, y: toFloat(x) / toFloat(y),
  1: lambda x, y: toFloat(x) * toFloat(y),
  2: lambda x, y: toFloat(x) + toFloat(y),
  3: lambda x, y: toFloat(x) - toFloat(y)
}

# Follows strict PEDMAS
def Evaluate(eq, solve = False, repl = False, guess = 10, shouldGuessImag = False, head=""):   
  if isFloat(eq):
    return toFloat(eq)
    
  if repl:
    #print(head + f"Standardizing eq: '{eq}'")
    if '`' in eq:
      raise ValueError("Incomplete expression")

    eq = eq.replace("~", "-").replace(r"\im", "j")
  
    eq = impMult(eq)
  
    # Change constants into numbers
    eq = eq.replace("\\pi", f"{math.pi}").replace("\\e", f"{math.e}")
    #print(head + f"   Eq is {eq} after replacing constants")

  eq = impMult(eq)
    
  if solve:
    return NewtonMethod(eq, guess, 8, shouldGuessImag)
  
  #print()
  #print(head + f"Evaluating \'{eq}\'")
  
  # P - Parenthetical function (trig)
  found = None
  while found := [(eq.find(func + "("), func) for func in ["sin", "cos", "tan", "sec", "csc", "cot"] if func in eq]:
    pos, func = min(found, key=lambda x: x[0])  # Get first match in string

    is_inverse = False
    if eq[pos + 3:pos + 8] == "^{-1}":
      is_inverse = True
      func = "a" + func
    
    contents = Between(eq, eq.index(func[1:] if is_inverse else func) + (8 if is_inverse else 3), "(", ")")

    #print(head + f"  Found trig function {func} with contents {contents}")
    eq = eq[:pos] + str(trigs[func](Evaluate(contents, head="  "))) + eq[pos + (10 if is_inverse else 5) + len(contents):]
    #eq = eq.replace(f"{found[0][1] + "^{-1}" if is_inverse else func}({contents})", 
    #                str(trigs[func](Evaluate(contents))))

  # P - Parenthetical function (log)
  loc = None
  while (loc := eq.find("log")) != -1:
    base = Between(eq, loc + 3, "{", "}")
    contents = Between(eq, loc + len(base) + 6, "(", ")")
    #print(head + f"  Found logarithm with base {base} and contents {contents}")
    
    # 01234....910........19
    # log_{base}(contents)
    # 4 + len(base)(4) + 2 + len(contents)(8) + 1 + 1
    # = 8 + len(base)(4) + len(contents)(8)

    eq = (eq[:loc]
          + str(cmath.log(Evaluate(contents, head="  "), Evaluate(base, head="  "))) 
          + eq[loc + 8 + len(base) + len(contents):])
    #eq = eq.replace(f"log_{{{base}}}({contents})", str(cmath.log(Evaluate(contents), Evaluate(base))))

  # P - Parenthesis
  while "(" in eq or "[" in eq:
    pair = ("(", ")") if "(" in eq else ("[", "]")
    contents = Between(eq, 0, pair[0], pair[1])
    #print(head + f"  Found {'parenthesis' if pair[0] == '(' else 'absolute value'} with contents {contents}")
    
    eq = eq.replace(f"{pair[0]}{contents}{pair[1]}", 
                    str(abs(Evaluate(contents, head="  ")) if pair[0] == "[" else Evaluate(contents, head="  "))
                       .replace(pair[0], "").replace(pair[1], "")
                   )

  # E - Exponents
  loc = None
  while (loc := eq.find("^")) != -1:
    exp = Between(eq, loc, "{", "}")
    
    i = 0
    base = eq[:loc]
    while not isFloat(base):
      i += 1
      base = eq[:loc][i:]
    
    #print(head + f"  Found exponent with base {base} and exponent {exp}")
    eq = eq[:loc - len(base)] + str(pow(toFloat(base), Evaluate(exp, head="  "))) + eq[3 + loc + len(exp):]
    #eq = eq.replace(base + "^{"+exp+"}",
    #                str(pow(toFloat(base), Evaluate(exp))))

  # E - Exponents (square root)
  loc = None
  while (loc := eq.find("\\sqrt")) != -1:
    nthroot = Between(eq, loc + 5, "{", "}")
    contents = Between(eq, loc + 5 + len(nthroot) + 2, "{", "}")
    #print(head + f"  Found a {nthroot} degree radical containing {contents}")
    
    eq = eq[:loc] + str(pow(Evaluate(contents, head="  "), 1 / Evaluate(nthroot, head="  "))) + eq[loc + len(nthroot) + len(contents) + 9:]
    #eq = eq.replace(r"\sqrt{"+nthroot+"}" + "{"+contents+"}", 
    #                str(pow(Evaluate(contents), 1 / Evaluate(nthroot))))

  # D - Division (Fractions have priority)
  loc = None
  while (loc := eq.find("\\frac")) != -1:
    numer = Between(eq, loc + 5, "{", "}")
    denom = Between(eq, loc + len(numer) + 7, "{", "}")
    #print(head + f"  Found a fraction with numerator {numer} and denominator {denom}")
    
    eq = eq[:loc] + str(Evaluate(numer, head="  ") / Evaluate(denom, head="  ")) + eq[loc + len(numer) + len(denom) + 9:]
    #eq = eq.replace(f"\\frac{{{numer}}}{{{denom}}}",
    #                str(Evaluate(numer) / Evaluate(denom)))

  # Permutations and Combinations
  loc = None
  while (loc := eq.find("\\comb")) != -1 or (loc := eq.find("\\perm")) != -1:
    op = r"*\comb*" if eq[loc + 1] == "c" else r"*\perm*"
    
    i = 0
    while not isFloat(eq[:loc - 1][i:]):
      i += 1
    n = eq[:loc][i:]
    
    i = len(eq)
    while not isFloat(eq[loc + 6:][:i]):
      i -= 1
    r = eq[loc + 5:][:i]
    
    eq = eq.replace(n + op + r, str(math.perm(int(n), int(r)) / (math.factorial(int(r)) if op[2] == "c" else 1)))

  # Reevaluate parenthesis because complex numbers
  while "(" in eq:
    contents = Between(eq, 0, "(", ")")
    #print(head + f"  Found parenthesis with contents {contents}")
    
    eq = eq.replace(f"({contents})", str(Evaluate(contents, head="  ")).replace("(", "").replace(")", ""))

  # DMAS - In that order
  curOp = 0
  progress = 0
  doNotCheck = False
  #print(head + f"  Finding operations in {eq}")
  
  #start = monotonic_ns()
  while doNotCheck or not isFloat(eq):
    if (progress := eq.find(("/", "*", "+", "-")[curOp], progress)) == -1:
      curOp += 1 if curOp < 3 else -1
      progress = 0
      doNotCheck = True
      continue
    
    doNotCheck = False

    lenEq = len(eq)

    i = 0
    befOp = eq[:progress]
    first = befOp
    while not isFloat(first) or first[0] == " ": 
      first = befOp[(i := i + 1):]
    
    aftOp = eq[progress + 1:]
    j = lenEq - progress - 1
    second = aftOp
    while not isFloat(second):
      second = aftOp[:(j := j - 1)]

    #print(head + f"    found operator {curOp} with first {first} and second {second}")
    #print(f"result: {str(operate[curOp](first, second))}")

    repl = str(operate[curOp](first, second))
    if repl[0] == "(":
      repl = repl[1:-1]

    eq = eq[:i] + repl + eq[lenEq - (lenEq - progress - 1) + j:]
    progress -= (i - len(repl) + 1)

    #print(f"eq: {eq}")
  
  #print(head + f"  operated eq in {(monotonic_ns() - start) / 1_000_000}ms")

  ans = toFloat(eq)
  #print(head + f"Finished evaluating; result: {ans}\n")
  return complex(round(complex(ans).real, 15), round(complex(ans).imag, 15)) if repl else ans

def primeFactors(n):
  i = 2
  factors = []
  while i * i <= n:
    if n % i:
      i += 1
    else:
      n //= i
      factors.append(i)
  if n > 1:
    factors.append(n)
  return factors
  
# Add implicit multiplication
def impMult(eq):
  i = 0
  while i < len(eq) - 1:
    if ((eq[i].isdigit() and eq[i + 1] == "\\") or # 5\pi
        (eq[i].isdigit() and eq[i + 1] == "(") or # 2(3 ...
        (eq[i].isdigit() and eq[i + 1].isalpha() and eq[i + 1] != "e" and eq[i + 1] != "j") or # 9sin(... but not 3e or 3j
        (eq[i].isalpha() and eq[i + 1].isdigit()) or # \pi3...
        (eq[i].isalpha() and eq[i + 1] == "\\") or # \pi\e...
        (eq[i] == "j" and eq[i + 1].isalpha()) or # jsin(...
        (eq[i] == ")" and eq[i + 1] == "\\") or # ...4)\pi
        (eq[i] == ")" and eq[i + 1] == "(") or # ...3)(6...
        (eq[i] == ")" and eq[i + 1].isdigit()) or # ...)7
        (eq[i] == ")" and eq[i + 1].isalpha()) or # ...)x...
        (eq[i] == "}" and eq[i + 1].isdigit())): # ...2}7
      eq = eq[:i + 1] + " * " + eq[i + 1:] # Insert multiplication
    i += 1
  return eq

# Finds string subset between char1 and char2, with nesting support
def Between(string, start, char1, char2):
  char1Indx = 0
  char1Cnt = 0
  char2Cnt = 0
  for i in range(start, len(string)):
    if string[i] == char1:
      if char1Cnt == 0:
        char1Indx = i
      
      char1Cnt += 1
    elif string[i] == char2:
      char2Cnt += 1
      
      if char1Cnt == char2Cnt:
        return string[char1Indx + 1 : i]

  raise ValueError(f"Unable to find contents between {char1} and {char2}")


def isFloat(string):
  if len(string) == 0 or string[0] == "+": return False

  try:
    toFloat(string)
    return True

  except ValueError:
    return False
    
def toFloat(string):
  cmplx = complex(string)
  return cmplx if cmplx.imag != 0 else cmplx.real
  
inc = 0.001
def NewtonMethod(eq, guess, accuracy, shouldGuessImag, alter = 1, epsilon = 0.00001, maxIter = 40):
  if shouldGuessImag:
    guess = guess + 1j

  for i in range(maxIter):
    if i == maxIter - 1:
      raise ValueError(f"No root found under {maxIter} iterations")
    
    inputEQ = eq.replace("x", str(guess))
    funcAtGuess = Evaluate(inputEQ)
    if abs(funcAtGuess.real + funcAtGuess.imag) < 2 * math.pow(10, -accuracy - 1):
      break
    
    derivAtGuess = (funcAtGuess - Evaluate(eq.replace("x", str(guess - inc)))) / inc
    if abs(derivAtGuess) < epsilon:
      raise ValueError(f"Function too flat... Possible asymptote. Try another guess. {derivAtGuess}")
    
    testGuess = (guess * derivAtGuess - funcAtGuess) / derivAtGuess
    if not shouldGuessImag and Evaluate(eq.replace("x", str(testGuess))).imag != 0:
      doubleDerivAtGuess = (derivAtGuess - ((funcAtGuess - Evaluate(eq.replace("x", str(guess - inc - inc)))) / inc)) / inc
      #print(f"{'Decrimenting' if doubleDerivAtGuess.real < 0 else 'Incrementing'} guess {guess} by {alter}")
      guess += -alter if doubleDerivAtGuess.real < 0 else alter
      continue
    
    guess = testGuess
    #print(f"New Guess: {guess}")
    
  
  return complex(round(complex(guess).real, accuracy), round(complex(guess).imag, accuracy))

File: Evaluator.cpp
#include <iostream>
#include <cmath>
#include <complex>
#include <string>
#include <vector>
#include <sstream>
#include <iomanip>
#include <map>
#include "LiTeX.cpp"

typedef unsigned char ubyte;
typedef signed char sbyte;
typedef std::complex<double> cmplx;

const cmplx NaN(std::nan("0"), std::nan("0"));
inline bool isNan(cmplx z) { return z != z; }

sbyte sgn(double val) {
    return (0 < val) - (val < 0);
}

inline std::string CmplxToStr(cmplx z, ubyte prec = 10) {
    return (z.imag() == 0 || z.real() == 0)
         ? (std::stringstream() << std::setprecision(prec) << (z.imag() == 0 ? z.real() : z.imag()) << (z.imag() == 0 ? "" : "i")).str()
         : ("(" + (std::stringstream() << std::setprecision(prec) << z.real()).str() + (sgn(z.imag()) >= 0 ? "+" : "") + (std::stringstream() << z.imag()).str() + "i)");
}

// PARSE COMPLEX ===================================================================== PARSE COMPLEX
cmplx ParseComplex(std::string str) {
    double real = 0;
    double imag = 0;
    ubyte start = 0;
    std::string reason = "";

PARSE_START:
    ubyte stop = str.size() - (str[str.size() - 1] == ')' ? 1 : 0);
    start += (str[0] == '(' && start == 0 ? 1 : 0);
    for (ubyte i = start; i < stop; i++) {
        if (i == start && str[i] == '-') continue;
        
        // We found the end of a number! (Or the string ended (or it's just part of sci notation))
        bool isEnd = i == stop - 1;
        if (str[i] == '-' || str[i] == '+' || isEnd) {
            if (str[i - 1] == 'e') {
                continue; // It's fine guys. Don't worry about the exponent
            }
            
            if (str[i - (isEnd ? 0 : 1)] == 'i') {
                if (imag != 0) {
                    reason = "Multiple imaginary parts found";
                    goto ERROR; // Woah, we already found imag part
                }

                imag = std::stod(str.substr(start, i - start + (isEnd ? 0 : -1)));
                break; // Let's get out of here
            }
            else {
                if (real != 0) {
                    reason = "Multiple real parts found";
                    goto ERROR; // Woah, we already found real part
                }

                real = std::stod(str.substr(start, i - start + (isEnd ? 1 : 0)));
                start = i + (str[i] == '+' ? 1 : 0);
                
                if (!(i == str.size() - 1))
                    goto PARSE_START; // Restart parsing cuz there still might be imag part
                break;
            }
        }

        if (!(isdigit(str[i]) || str[i] == 'e' || str[i] == '.')) {
            reason = "Unrecongized character " + std::string(1, str[i]);
            goto ERROR;
        }
    }

    if (real != 0 && imag != 0 && 
        (str[0] != '(' && str[str.size() - 1] != ')')
    ) {
        reason = "Complex numbers must be surrounded by parenthesis";
        goto ERROR;
    }

    return cmplx(real, imag);
    
ERROR:
    std::cerr << "                                                    >>> COULD NOT PARSE '" << str << "' FOR COMPLEX (" << reason << ")" << std::endl;
    return NaN;
}

// FIND COMPLEX ====================================================================== FIND COMPLEX
// Avoid slicing and checking with ParseComplex, which is slow
// Returns complex-parseable string.
std::string FindCmplx(std::string str, ubyte loc, bool before) {
    ubyte minusCount = 0;
    ubyte plusCount = 0;
    for (ubyte i = loc + (before ? -1 : 1); 
        before ? (i >= 0) : (i < str.size()); 
        i += (before ? -1 : 1)
    ) {
        if (i == loc - 1 && before && str[i] == ')') {
            // Found true complex, return contents
            return "(" + Between(str, i, ')', '(', true) + ")";
        }
        else if (!before && i == loc + 1 && str[i] == '(')
            return "(" + Between(str, i, '(', ')') + ")";

        if (str[i] == '+') // Must be part of exp
            if (str[i - 1] != 'e' || plusCount != 0)
                goto ERROR;
            else 
                plusCount++;
        
        else if (str[i] == '-') { // Found minus, could be negative or e-...
            if (!before) {
                if (i == loc + 1) {
                    minusCount++;
                }
                else if (str[i - 1] != 'e' || minusCount > 1)
                    return str.substr(loc + 1, i - loc - 1);
            }
            else {
                if (str[i - 1] == 'e' && minusCount == 0)
                minusCount++; // Increase minus count
                else if (minusCount == 1) // Negative number
                    return str.substr(i, loc - i);
                else
                    goto ERROR;
            }
        }
        else if (str[i] == 'e' && (!(str[i + 1] == '+' || str[i + 1] == '-') || !isdigit(str[i + 2])))
            goto ERROR;
        // Encountered unidentified character, must be end of number
        else if (!(isdigit(str[i]) || str[i] == 'e' || str[i] == '.' || (str[i] == 'i' && before && i == loc - 1)) || i == (before ? 0 : str.size() - 1))
            return before ? str.substr(i, loc - i) : str.substr(loc + 1, i - loc);
    }

ERROR:
    std::cerr << "                                                    >>> COULD NOT FIND COMPLEX" << std::endl;
    return "";
}

// ADD IMP MULT ====================================================================== ADD IMP MULT
void AddImpMultTo(std::string& eq) {
    for (ubyte i = 1; i < eq.size() - 1; i++) {
        if ((isdigit(eq[i]) && eq[i + 1] == '\\') || // 5\pi
            (isdigit(eq[i]) && eq[i + 1] == '(') || // 2(3...
            (isdigit(eq[i]) && isalpha(eq[i + 1]) 
                && eq[i + 1] != 'e' && eq[i + 1] != 'i') || // 9sin... but not 3e+... or 3i...
            (isalpha(eq[i]) && isdigit(eq[i + 1])) || // \pi3...
            (isalpha(eq[i]) && eq[i + 1] == '\\') || // \pi\e...
            (eq[i] == ')' && eq[i + 1] == '\\') || // ...4)\pi
            (eq[i] == ')' && eq[i + 1] == '(') || // ...3)(6...
            (eq[i] == ')' && isdigit(eq[i + 1])) || // ...)7
            (eq[i] == ')' && isalpha(eq[i + 1])) || // ...)x...
            (eq[i] == '}' && isdigit(eq[i + 1]))) // ...2}7
        eq.insert(i + 1, 1, '*');
    }
}

const std::unordered_map<std::string, ubyte> trigMap = {
    { "sin", 1 },
    { "cos", 2 },
    { "tan", 3 },
    { "sec", 4 },
    { "csc", 5 },
    { "cot", 6 },
    { "sin^{-1}", 7 },
    { "cos^{-1}", 8 },
    { "tan^{-1}", 9 },
};
auto operateTrig = [](std::string func, cmplx n) {
    switch (trigMap.at(func)) {
        case 1: return sin(n);
        case 2: return cos(n);
        case 3: return tan(n);
        case 4: return cmplx(1,0) / sin(n);
        case 5: return cmplx(1,0) / cos(n);
        case 6: return cmplx(1,0) / tan(n);
        case 7: return asin(n);
        case 8: return acos(n);
        case 9: return atan(n);
    }
    return NaN;
};
auto operate = [](ubyte op, cmplx a, cmplx b) {
    switch (op) {
        case 0: return a / b;
        case 1: return a * b;
        case 2: return a - b;
        case 3: return a + b;
    }
    return NaN;
};

// FIND TRIGS ======================================================================== FIND TRIGS
std::tuple<std::string, ubyte> FindTrigs(std::string eq, std::array<std::string, 9> trigs) {
    for (ubyte i = 0; i < size(eq); i++) {
        for (ubyte n = 0; n < trigs.size(); n++) {
            if (eq.substr(i, trigs[n].size()) == trigs[n])
                return std::tuple<std::string, ubyte>(trigs[n], i);
        }
    }
    std::cerr << "                                                    >>> COULD NOT FIND TRIG IN EQ: '" << eq << "'" << std::endl;
    return std::tuple<std::string, ubyte>("", 0xFF);
}

// EVALUATE ========================================================================== EVALUATE
cmplx Evaluate(std::string eq) {
    std::string reason = "";
{ // Scope so goto ERROR will not bypass variable initialization
    
    // SETUP EQUATION ==================================================== SETUP EQUATION
    {   cmplx test; // Is eq a number? (Scope so test is del)
        std::cout << "Testing whether eq is complex..." << std::endl;
        if (!isNan(test = ParseComplex(eq))) return test;   }

    AddImpMultTo(eq);
    
    // ` is placeholder character for renderer.
    if (eq.find('`') != std::string::npos) { reason = "Equation has placeholders"; goto ERROR; }
    
 /* TRIGONOMETRIC FUNCTIONS */ {
        std::cout << "Finding trig functions..." << std::endl;
        std::tuple<std::string, ubyte> loc;
        while((loc = FindTrigs(eq, std::array<std::string, 9>({ "sin", "cos", "tan", "sec", "csc", "cot", "sin^{-1}", "cos^{-1}", "tan^{-1}" })))
            != std::tuple<std::string, ubyte>("", 0xFF)
        ) {
            std::cout << "Found trig function " << std::get<0>(loc) << std::endl;
            std::string contents = Between(eq, std::get<1>(loc) + std::get<0>(loc).size(), '(', ')');
            
            eq.erase(std::get<1>(loc), std::get<0>(loc).size() + contents.size() + 2);
            eq.insert(std::get<1>(loc), CmplxToStr(operateTrig(std::get<0>(loc), Evaluate(contents))));
        }
    } // Scope so loc is freed. (I want to change type of loc from tuple -> size_t)

    // LOGARITHMS
    size_t loc = -1;
    while((loc = eq.find('log')) != std::string::npos) {
        std::string base = Between(eq, loc + 4, '{', '}');
        std::string contents = Between(eq, loc + base.size() + 6, '(', ')');

        eq.erase(loc, base.size() + contents.size() + 8);
        eq.insert(loc, CmplxToStr(log(Evaluate(contents)) / log(Evaluate(base))));
        // Use change-of-base formula (inefficient, but too lazy to write my own variable base thing)
    }

    // PARENTHESIS
    loc = -1;
    while ((loc = eq.find('(', loc + 1)) != std::string::npos) {
        std::string contents = Between(eq, loc, '(', ')');
        std::cout << "Found parenthesis with contents: " << contents << std::endl;
        
        cmplx toRepl = ParseComplex("(" + contents + ")");
        if (isNan(toRepl)) toRepl = Evaluate(contents);

        eq.erase(loc, contents.size() + 2);
        eq.insert(loc, CmplxToStr(toRepl));
    }
    
    // RADICALS
    loc = -1;
    while ((loc = eq.find("\\sqrt", loc + 1)) != std::string::npos) {
        std::string nth = Between(eq, loc + 5, '{', '}');
        std::string root = Between(eq, loc + 7 + nth.size(), '{', '}');

        eq.erase(loc, nth.size() + 7 + root.size());
        eq.insert(loc, CmplxToStr(pow(Evaluate(root), cmplx(1, 0) / Evaluate(nth))));
    }

    // FRACTIONS
    loc = -1;
    while ((loc = eq.find("\\frac", loc + 1)) != std::string::npos) {
        std::string numer = Between(eq, loc + 5, '{', '}');
        std::string denom = Between(eq, loc + 7 + numer.size(), '{', '}');

        eq.erase(loc, numer.size() + 7 + denom.size());
        eq.insert(loc, CmplxToStr(Evaluate(numer) / Evaluate(denom)));
    }

    // EXPONENTS
    loc = -1;
    while ((loc = eq.find('^', loc + 1)) != std::string::npos) {
        std::string exp = Between(eq, loc + 1, '{', '}');
        std::string base = FindCmplx(eq, loc, true);
        std::cout << "Found exponent with base " << base << " and exponent " << exp << std::endl;

        eq.erase(loc - base.size(), base.size() + 3 + exp.size());
        eq.insert(loc - base.size(), 
            CmplxToStr(std::pow(ParseComplex(base), Evaluate(exp))));
    }

    // OPERATORS ========================================================= OPERATORS
    std::cout << "Finding operators..." << std::endl;
    ubyte curOp = 0;
    size_t progress = 0;
    bool DNC = false;
    const char ops[] = { '/', '*', '-', '+' };
    while (DNC || isNan(ParseComplex(eq))) {
        if ((progress = eq.find(ops[curOp]), progress) == std::string::npos) {
            curOp++;
            progress = 0;
            DNC = true;
            continue;
        }
        DNC = false;

        std::string first = FindCmplx(eq, progress, true);
        std::string second = FindCmplx(eq, progress, false);
        std::cout << "Found operator " << ops[curOp] << " with first " << first << " and second " << second << std::endl;

        std::cout << "Old eq: " << eq;
        eq.erase(progress - first.size(), first.size() + 1 + second.size());
        std::cout << "; Replacing operation, new eq: '" << eq << '\'' << std::endl;

        std::string toReplace = CmplxToStr(operate(curOp, ParseComplex(first), ParseComplex(second)));
        std::cout << "Result of operation: " << CmplxToStr(operate(curOp, ParseComplex(first), ParseComplex(second))) << std::endl;

        eq.insert(progress - first.size(), toReplace);
        progress += (toReplace.size() - first.size()) + 1;
    }   

    return ParseComplex(eq);
} // End error scope (Not evaluate)

ERROR:
    std::cerr << "                                                    >>> COULD NOT EVALUATE EQ'" << eq << "' (" << reason << ")" << std::endl;
    return NaN;
} // End evaluate